app/controllers/actions/IdentifierAction.scala

```diff
@@ -17,69 +17,180 @@
+import models.requests.{IdentifiedRequest, IdentityData}
+import models.{NormalMode, SignedInUser}
+import play.api.Logger
 import play.api.mvc.Results._
 import play.api.mvc._
 import uk.gov.hmrc.auth.core._
-import uk.gov.hmrc.auth.core.retrieve.v2.Retrievals
-import uk.gov.hmrc.http.{HeaderCarrier, UnauthorizedException}
+import uk.gov.hmrc.auth.core.authorise.Predicate
+import uk.gov.hmrc.auth.core.retrieve.~
+import uk.gov.hmrc.http.HeaderCarrier
 import uk.gov.hmrc.play.http.HeaderCarrierConverter
+import controllers.home.{routes => homeRoutes}
+import controllers.agents.{routes => agentRoutes}
 
 import scala.concurrent.{ExecutionContext, Future}
 
+object IdentifierAction {
+  val pptEnrolmentKey            = "HMRC-PPT-ORG"
+  val pptEnrolmentIdentifierName = "EtmpRegistrationNumber"
+}
+
 trait IdentifierAction
-    extends ActionBuilder[IdentifierRequest, AnyContent]
-    with ActionFunction[Request, IdentifierRequest]
+    extends ActionBuilder[IdentifiedRequest, AnyContent]
+    with ActionFunction[Request, IdentifiedRequest]
 
 class AuthenticatedIdentifierAction @Inject() (
   override val authConnector: AuthConnector,
-  config: FrontendAppConfig,
+  pptReferenceAllowedList: PptReferenceAllowedList,
+  override val appConfig: FrontendAppConfig,
+  metrics: Metrics,
   val parser: BodyParsers.Default
 )(implicit val executionContext: ExecutionContext)
-    extends IdentifierAction with AuthorisedFunctions {
+    extends IdentifierAction with AuthorisedFunctions with CommonAuth {
+
+  private val authTimer = metrics.defaultRegistry.timer("ppt.returns.upstream.auth.timer")
+  private val logger    = Logger(this.getClass)
 
   override def invokeBlock[A](
     request: Request[A],
-    block: IdentifierRequest[A] => Future[Result]
+    block: IdentifiedRequest[A] => Future[Result]
   ): Future[Result] = {
 
     implicit val hc: HeaderCarrier =
       HeaderCarrierConverter.fromRequestAndSession(request, request.session)
 
-    authorised().retrieve(Retrievals.internalId) {
-      _.map {
-        internalId => block(IdentifierRequest(request, internalId))
-      }.getOrElse(throw new UnauthorizedException("Unable to retrieve internal Id"))
+    // In theory this could be deduplicated with SelectedClientIdentifier by the generic makes it too difficult
+    def getSelectedClientIdentifier() = request.session.get("clientPPT")
+
+    val authorisation            = authTimer.time()
+    val selectedClientIdentifier = getSelectedClientIdentifier()
+
+    authorised(authPredicate(selectedClientIdentifier)).retrieve(authData) {
+      case credentials ~ name ~ email ~ externalId ~ internalId ~ affinityGroup ~ allEnrolments ~ agentCode ~
+          confidenceLevel ~ authNino ~ saUtr ~ dateOfBirth ~ agentInformation ~ groupIdentifier ~
+          credentialRole ~ mdtpInformation ~ itmpName ~ itmpDateOfBirth ~ itmpAddress ~ credentialStrength ~ loginTimes =>
+        authorisation.stop()
+
+        val maybeInternalId = internalId.getOrElse(
```

Comment:
this name is confusing i thin, maybe normally refers to an option

Reply:
Changed